client/ws-client/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+
+#[cfg(test)]
+mod tests;
+
+pub use jsonrpsee_types as types;
+
+use jsonrpsee_client_transport::ws::{Header, InvalidUri, Uri, WsTransportClientBuilder};
+use jsonrpsee_core_client::{Client, ClientBuilder};
+use std::time::Duration;
+use types::{CertificateStore, Error, TEN_MB_SIZE_BYTES};
+
+/// Builder for [`Client`].
+///
+/// # Examples
+///
+/// ```no_run
+///
+/// use jsonrpsee_ws_client::WsClientBuilder;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     // build client
+///     let client = WsClientBuilder::default()
+///          .add_header("Any-Header-You-Like", "42")
+///          .build("wss://localhost:443")
+///          .await
+///          .unwrap();
+///
+///     // use client....
+/// }
+///
+/// ```
+#[derive(Clone, Debug)]
+pub struct WsClientBuilder<'a> {
+	certificate_store: CertificateStore,
+	max_request_body_size: u32,
+	request_timeout: Duration,
+	connection_timeout: Duration,
+	headers: Vec<Header<'a>>,
+	max_concurrent_requests: usize,
+	max_notifs_per_subscription: usize,
+	max_redirections: usize,
+}
+
+impl<'a> Default for WsClientBuilder<'a> {
+	fn default() -> Self {
+		Self {
+			certificate_store: CertificateStore::Native,
+			max_request_body_size: TEN_MB_SIZE_BYTES,
+			request_timeout: Duration::from_secs(60),
+			connection_timeout: Duration::from_secs(10),
+			headers: Vec::new(),
+			max_concurrent_requests: 256,
+			max_notifs_per_subscription: 1024,
+			max_redirections: 5,
+		}
+	}
+}
+
+impl<'a> WsClientBuilder<'a> {
+	/// Set whether to use system certificates
+	pub fn certificate_store(mut self, certificate_store: CertificateStore) -> Self {
+		self.certificate_store = certificate_store;
+		self
+	}
+
+	/// Set max request body size.
+	pub fn max_request_body_size(mut self, size: u32) -> Self {
+		self.max_request_body_size = size;
+		self
+	}
+
+	/// Set request timeout (default is 60 seconds).
+	pub fn request_timeout(mut self, timeout: Duration) -> Self {
+		self.request_timeout = timeout;
+		self
+	}
+
+	/// Set connection timeout for the handshake.
+	pub fn connection_timeout(mut self, timeout: Duration) -> Self {
+		self.connection_timeout = timeout;
+		self
+	}
+
+	/// Set a custom header passed to the server during the handshake.
+	///
+	/// The caller is responsible for checking that the headers do not conflict or are duplicated.
+	pub fn add_header(mut self, name: &'a str, value: &'a str) -> Self {
+		self.headers.push(Header { name, value: value.as_bytes() });
+		self
+	}
+
+	/// Set max concurrent requests.
+	pub fn max_concurrent_requests(mut self, max: usize) -> Self {
+		self.max_concurrent_requests = max;
+		self
+	}
+
+	/// Set max concurrent notification capacity for each subscription; when the capacity is exceeded the subscription
+	/// will be dropped.
+	///
+	/// You can also prevent the subscription being dropped by calling
+	/// [`Subscription::next()`](crate::types::Subscription) frequently enough such that the buffer capacity doesn't
+	/// exceeds.
```

Comment:
```suggestion
	/// exceed its limit.
```

Im a little confused by `frequently enough`. Does this mean a user should monitor the subscriptions and as it starts to get close to exceeding its capacity to call `[`Subscription::next()`]`, therefore the implementation on what `frequently` is, is based on the user?

Reply:
Maybe it's poorly phrased but it means that if the server produces items much faster than you would call `Subscription::next()` that might end up with the filling buffer and once it exceeds the configured `max_limit`.

In that scenario you need to either call/poll `Subscription::next` more often or increase the buffer max limit, k?

Reply:
Ahh, okay yea that totally makes sense. 

> more often or increase the buffer max limit

This is what i assumed, just needed some clarification :). Thanks for the explanation